procrastinate/retry.py

```diff
@@ -6,28 +6,88 @@
 from __future__ import annotations
 
 import datetime
-from typing import Iterable, Union
+import warnings
+from typing import Iterable, Union, overload
 
 import attr
 
 from procrastinate import exceptions, utils
+from procrastinate.jobs import Job
+
+
+class RetryDecision:
+    retry_at: datetime.datetime | None = None
+    priority: int | None = None
+    queue: str | None = None
+    lock: str | None = None
+
+    @overload
+    def __init__(
+        self,
+        retry_at: None = None,
+        retry_in: utils.TimeDeltaParams | None = None,
+        priority: int | None = None,
+        queue: str | None = None,
+        lock: str | None = None,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        retry_at: datetime.datetime | None = None,
+        retry_in: None = None,
```

Comment:
You can remove parameters entirely from overloads if it helps. Here I believe you may remove retry_at from the top one and retry_in from the bottom one

Reply:
I thought so too, but then I ended up with the below type error (at least in VS Code with pyright extension):
```
Overloaded implementation is not consistent with signature of overload 1
  Type "(self: Self@RetryDecision, retry_at: datetime | None = None, retry_in: TimeDeltaParams | None = None, priority: int | None = None, queue: str | None = None, lock: str | None = None) -> None" is incompatible with type "(self: Self@RetryDecision, retry_in: TimeDeltaParams | None = None, priority: int | None = None, queue: str | None = None, lock: str | None = None) -> None"
    Parameter name mismatch: "retry_in" versus "retry_at"
    Parameter 2: type "TimeDeltaParams | None" is incompatible with type "datetime | None"
    Parameter name mismatch: "priority" versus "retry_in"
    Parameter 3: type "int | None" is incompatible with type "TimeDeltaParams | None"
    Parameter name mismatch: "queue" versus "priority"
    Parameter 4: type "str | None" is incompatible with type "int | None"
    Parameter name mismatch: "lock" versus "queue"Pyright[reportInconsistentOverload](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportInconsistentOverload)
retry.py(25, 9): Overload signature is defined here
Overloaded implementation is not consistent with signature of overload 2
  Type "(self: Self@RetryDecision, retry_at: datetime | None = None, retry_in: TimeDeltaParams | None = None, priority: int | None = None, queue: str | None = None, lock: str | None = None) -> None" is incompatible with type "(self: Self@RetryDecision, retry_at: datetime | None = None, priority: int | None = None, queue: str | None = None, lock: str | None = None) -> None"
    Parameter name mismatch: "priority" versus "retry_in"
    Parameter 3: type "int | None" is incompatible with type "TimeDeltaParams | None"
    Parameter name mismatch: "queue" versus "priority"
    Parameter 4: type "str | None" is incompatible with type "int | None"
    Parameter name mismatch: "lock" versus "queue"Pyright[reportInconsistentOverload](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportInconsistentOverload)
retry.py(33, 9): Overload signature is defined here
```

Reply:
Can you try adding a `*` to make those kw only ?

Reply:
Sure, I haven't thought of that. This works fine 😌.